### PR TITLE
Add parade: a PostgreSQL-based task queue

### PR DESCRIPTION
**Pull request by arielshaqed**
Instructions for reviewer:

1. Please read #609 to understand what it's all about.  This PR implements _most_ but not _all_ of that spec.  In particular, `ExtendTask` will be added later.
1. Now read the SQL.  It's pretty much self-contained and has almost all of the business logic.  You may want to read about `SELECT ... FOR UPDATE SKIP UNLOCKED`; good references are [the very fine manual](https://www.postgresql.org/docs/12/sql-select.html#SQL-FOR-UPDATE-SHARE) and a [What is SKIP LOCKED for in PostgreSQL 9.5?](https://www.2ndquadrant.com/en/blog/what-is-select-skip-locked-for-in-postgresql-9-5/).
1. Now read `ddl.go` -- mostly a direct access layer to the SQL.  It _does_ use `pgx` (rather than `sql` or `sqlx`) for some operations that are not supported on `sql*`.  Question for reviewer: should it use *only* `pgx`?
1. Finally read `ddl_test.go`, which actually uses it.

### CODE REVIEW

**Comment by arielshaqed** [comment]
@guy-har yours for review, thanks!  (And *very* happy for any other reviewers, this needs any scrutiny it can get...)

**Comment by arielshaqed** [comment]
Now ready for @tzahij too, I think :-)

**Review by arielshaqed — COMMENTED** [review]
Thanks!

PTAL.

**Review by guy-har — APPROVED** [review]
LGTM 
But I'd prefer that another reviewer with SQL experience (maybe @tzahij ) would also take a look and approve. at least at the SQL part.

**Comment by arielshaqed** [comment]
> LGTM
> But I'd prefer that another reviewer with SQL experience (maybe @tzahij ) would also take a look and approve. at least at the SQL part.

@tzahij and/or @nopcoder: could either of you, please?

File: parade/ddl.sql [hunk 1]
```diff
@@ -0,0 +1,160 @@
+CREATE EXTENSION pgcrypto;
+
+CREATE SCHEMA IF NOT EXISTS parade;
+
+-- (requires extensioon pgcrypto)
```
Comment (nopcoder) [inline parade/ddl.sql:5]: extension

File: parade/ddl.sql [hunk 2]
```diff
@@ -0,0 +1,160 @@
+CREATE EXTENSION pgcrypto;
```
Comment (nopcoder) [inline parade/ddl.sql:1]: I remember something about extention and postgres that will require to check and install once on public

File: parade/ddl.sql [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+CREATE SCHEMA IF NOT EXISTS parade;
+
+-- (requires extensioon pgcrypto)
+
+CREATE TYPE task_status_code_value AS ENUM (
+    'pending',		-- waiting for an actor to perform it (new or being retried)
+    'in-progress',	-- task is being performed by an actor
```
Comment (nopcoder) [inline parade/ddl.sql:9]: indent comment

File: parade/ddl.sql (hunk 2 above)
Comment (arielshaqed) [reply parade/ddl.sql]: Not sure what you are requesting.  Up until now I've not gone through the up/down scripts, so it was a skipped error.

Now I've changed it to "if not exists", so it should always succeed.  `pgcrypto` is a standard extension, it should always be there.  And it _is_ there on RDS

File: parade/ddl.sql (hunk 1 above)
Comment (arielshaqed) [reply parade/ddl.sql]: Removed that comment, and the schema: we don't need either of them.

File: parade/ddl.sql (hunk 3 above)
Comment (arielshaqed) [reply parade/ddl.sql]: Done, untabified everything!  (Tabs vs. spaces, grr...).

**Review by arielshaqed — COMMENTED** [review]
Thanks; PTAL!

**Review by nopcoder — APPROVED** [review]

File: parade/ddl.sql [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+);
+
+CREATE TABLE IF NOT EXISTS tasks (
+    id VARCHAR(64) NOT NULL PRIMARY KEY, -- nanoid
+
+    action VARCHAR(128) NOT NULL, -- name (type) of action to perform
+    body TEXT,			-- data used by action
+    status TEXT,		-- status text defined by action, visible to action
+
+    status_code TASK_STATUS_CODE_VALUE NOT NULL DEFAULT 'pending', -- internal status code, used by parade to issue tasks
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply parade/ddl.sql]: It's the type defined at [line 3](https://github.com/treeverse/lakeFS/pull/641/files#diff-56bcbaf350e94cb9e8d599800f450846R3).  Changed to lowercase; we don't have an SQL style, but clearly here lowercase would be less confusing to at least one reviewer :-)

File: parade/ddl.sql [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    num_tries INTEGER NOT NULL DEFAULT 0,
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply parade/ddl.sql]: Assuming you meant a _comment_, and adding that.

File: parade/ddl.sql [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    num_signals INTEGER NOT NULL DEFAULT 0, -- number of tasks that have already signalled this task
+
+    -- BUG(ariels): add REFERENCES dependency to each of the to_signal
+    --     tasks.  Or at least add triggers that perform ON DELETE
+    --     CASCADE.
+    to_signal VARCHAR(64) ARRAY, -- IDs to signal after performing this task
+
+    actor_id VARCHAR(64),    -- ID of performing actor if in-progress
+    action_deadline TIMESTAMPTZ, -- offer this task to other actors once action_deadline has elapsed
+    performance_token UUID,
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply parade/ddl.sql:32]: https://github.com/treeverse/lakeFS/blob/master/design/task-management-api.md#life-of-a-task

File: parade/ddl.sql [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    finish_channel VARCHAR(64) -- (if non-NULL) name of a channel to NOTIFY when this task ends
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply parade/ddl.sql:33]: This is just a field on a task.  Will rename them all in one go, after I commit the entire stack of PRs that have accumulated.  Please add your suggestions for naming on #773.

File: parade/ddl.sql [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+-- TODO(ariels): add a lock token to each row, set when leasing the
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply parade/ddl.sql]: Naah, thanks!

File: parade/ddl.sql [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    SET status = result_status,
+    	status_code = result_status_code,
+	actor_id = NULL,
+	performance_token = NULL
+    WHERE id = task_id AND performance_token = token
+    RETURNING finish_channel, to_signal;
+
+    GET DIAGNOSTICS num_updated := ROW_COUNT;
+
+    IF channel IS NOT NULL THEN
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply parade/ddl.sql:99]: It doesn't particularly matter, there's either a transaction around it (and then notify happens last) or there isn't (and then notify is inherently async).  So at the very least you're right and it  is clearer to move it last!

**Review by arielshaqed — COMMENTED** [review]
Thanks; PTAL!

**Comment by arielshaqed** [comment]
Thanks!

Pulling...

**Pull request merged**